python/cudf/cudf/utils/cudautils.py

```diff
@@ -553,21 +553,27 @@ def gpu_unique_set_insert(vset, sz, val):
 
 
 @cuda.jit
-def gpu_shift(in_col, out_col, N):
+def gpu_shift(in_col, out_col, N, mask_bits):
     """Shift value at index i of an input array forward by N positions and
     store the output in a new array.
     """
     i = cuda.grid(1)
+
     if N > 0:
         if i < in_col.size:
             out_col[i] = in_col[i - N]
-        if i < N:
-            out_col[i] = -1
     else:
         if i <= (in_col.size + N):
             out_col[i] = in_col[i - N]
-        if i >= (in_col.size + N) and i < in_col.size:
-            out_col[i] = -1
+
+    base = i * mask_bitsize
+    for k in range(base, base + mask_bitsize):
+        if k >= in_col.size:
+            break
+        if N > 0 and k >= N:
+            mask_set(mask_bits, k)
```

Comment:
I don't think there's a need to set individual bits here (which invokes a global atomic per bit---expensive).

We're updating contiguous ranges of bits, so instead you can write whole bitmask elements at a time.

Something like this:
```
warp_size = 32
warp_id = i / warp_size
lane_id = i % warp_size
if( warp_id < N / warp_size)
   if(lane_id == 0)
      mask_bits[warp_id] = 0 # zero out the contiguous range of bits
remainder = N % warp_size
if( remainder != 0 and warp_id == (N/warp_size + 1) )
   new_mask = ~(1 << remainder) - 1 # mask with "remainder" LSBs cleared
   mask_bits[warp_id] &= new_mask # bitwise AND with old mask 
```

Reply:
@galipremsagar so `warp_size` would be `cuda.warpsize` and `lane_id` would be `cuda.laneid` per here: https://numba.pydata.org/numba-doc/dev/cuda-reference/kernel.html?highlight=warp#numba.cuda.warpsize